SL-917: add require expression functions for appframework configuration

cioan opened the pull request:
@mseaton , please let me know if there is a better place to store those java script functions. Thanks!

mseaton commented on configuration/appframework/scripts/isPatientEnrolledInProgramOfUuid.js:
```diff
@@ -0,0 +1,6 @@
+function isPatientEnrolledInProgramOfUuid(programUuid) {
```
Let's name this more concisely - we don't need `InProgramOfUuid` in the file name or method name

mseaton commented on configuration/appframework/scripts/patientHasEncounterOfTypeDuringProgramEnrollment.js:
```diff
@@ -0,0 +1,8 @@
+  if ( !isPatientEnrolled ) {
+    return false;
+  }
+  return isPatientEnrolledInProgramOfUuid(programUuid) && encounterOfTypeAfterDate(encounters, encounterTypeUuid, getDateEnrolledByProgramUuid(activePrograms, programUuid));
```
You don't need to evaluate `isPatientEnrolledInProgamOfUuid` again here, you already did it above.  Can't we just have this final statement?

cioan commented:
Thanks @mseaton ! I have updated the PR based on your review.

mseaton commented on configuration/appframework/scripts/getDateEnrolledByProgramUuid.js:
```diff
@@ -0,0 +1,12 @@
+    return false;
+  }
+  var i, len = patientPrograms.length;
+  for (i = 0; i < len; ++i) {
```
See comment above - use `patientPrograms.forEach`, rather than all of these variables.

mseaton commented on configuration/appframework/scripts/getDateEnrolledByProgramUuid.js:
```diff
@@ -0,0 +1,12 @@
+      return patientPrograms[i]['dateEnrolled'];
+    }
+  }
+  return false;
```
Same as above.  Why are you returning a boolean here?

mseaton commented on configuration/appframework/scripts/isPatientEnrolledInProgramOfUuid.js:
```diff
@@ -0,0 +1,6 @@
+function isPatientEnrolled(programUuid) {
```
You need to change this file name to match the function name, now that you have changed it.  Also, this is inconsistent with the way you have implemented the other functions.  The way you have implemented the others is to have all of the arguments you need passed into the function, whereas this one assumes some arguments are already defined as global variables.  We should choose one of these approaches and stick to it for consistency.

mseaton commented:
See my new comments @cioan

cioan commented on configuration/appframework/scripts/encounterOfTypeAfterDate.js:
```diff
@@ -0,0 +1,17 @@
+  }
+
+  var i, len = encounters.length;
+  for (i = 0; i < len; ++i) {
```
@mseaton, sure, but is this javascript syntax supported by the engine? I was trying to avoid any newer methods and syntax.

mseaton replied:
@cioan you might be right but might as well check?  If it isn't supported, you could use the `some` function like you do in the `isPatientEnrolled` function, right?

cioan commented:
Thanks @mseaton ! I think I addressed all your comments above, except .each() which I did not get a chance to test to see if it's supported by the engine.

mseaton commented on configuration/appframework/scripts/getDateEnrolledByProgramUuid.js:
```diff
@@ -0,0 +1,12 @@
+  var i, len = patientPrograms.length;
+  for (i = 0; i < len; ++i) {
+    if (patientPrograms[i]['programUuid'] === programUuid) {
+      return patientPrograms[i]['dateEnrolled'];
```
What if the patient has multiple enrollments for the same program?  What happens then?  Could easily happen in the pregnancy program.  What is the use case for this function and can we make it more explicit?

mseaton commented on configuration/appframework/scripts/isPatientEnrolled.js:
```diff
@@ -0,0 +1,6 @@
+function isPatientEnrolled(programs, programUuid) {
```
Based on your logic here, I'd name this something like `isEverEnrolled`.  If this is intended to indicate if there is a _current_ / _active_ enrollment, then that is not being checked here.  `isPatientEnrolled` implies that they are currently enrolled.

mseaton commented on configuration/appframework/scripts/patientHasEncounterOfTypeDuringProgramEnrollment.js:
```diff
@@ -0,0 +1,12 @@
+  var dateEnrolled = getDateEnrolledByProgramUuid(programs, programUuid);
+  if ( !dateEnrolled ) {
+    return false;
+  }
```
I don't see the point in checking both of these things above.  All you need is the `dateEnrolled`.  No point in checking `isPatientEnrolled` when a null `dateEnrolled` gives you the same information.

cioan commented:
@mseaton , I found it to be much clearer to just have one function that checks if patient has an encounter of given type during a given program enrollment. I have removed all other script functions. Now, since I have the entire logic in one function, and tested successfully with multiple patient visits and program enrollments, I am happy to break it down in smaller functions if you think that it is still necessary?

mseaton commented on configuration/appframework/scripts/patientHasEncounterOfTypeDuringProgramEnrollment.js:
```diff
@@ -0,0 +1,29 @@
+
+  var foundProgram = null;
+  for ( var i = 0; i < patientPrograms.length ; ++i ) {
+    if (( patientPrograms[i].programUuid === programUuid ) && visit
```
It seems that `visit` is required for this to work, so shouldn't we move the `&& visit` to line 3 with the other checks?

cioan commented on configuration/appframework/scripts/patientHasEncounterOfTypeDuringProgramEnrollment.js:
```diff
@@ -0,0 +1,29 @@
+  for ( var i = 0; i < patientPrograms.length ; ++i ) {
+    if (( patientPrograms[i].programUuid === programUuid ) && visit
+      && ( (new Date(visit.startDatetimeInMilliseconds)) >= ( new Date(patientPrograms[i].dateEnrolled)))
+      && ( patientPrograms[i].dateCompleted == null || (new Date(patientPrograms[i].dateCompleted)) >= (new Date(visit.startDatetimeInMilliseconds))) ) {
```
Thanks @mseaton ! I have updated the PR with your suggestion.
